src/server.rs

```diff
@@ -0,0 +1,153 @@
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl Offers for LNDKServer {
+    async fn pay_offer(
+        &self,
+        request: Request<PayOfferRequest>,
+    ) -> Result<Response<PayOfferResponse>, Status> {
+        log::info!("Received a request: {:?}", request);
+
+        let metadata = request.metadata();
+        let inner_request = request.get_ref();
+
+        let lnd_cfg = check_auth_metadata(metadata)?;
+        let mut client = get_lnd_client(lnd_cfg)
+            .map_err(|e| Status::unavailable(format!("Couldn't connect to lnd: {e}")))?;
+
+        let offer = Offer::from_str(&inner_request.offer).map_err(|e| {
+            Status::invalid_argument(format!(
+                "The provided offer was invalid. Please provide a valid offer in bech32 format,
+                i.e. starting with 'lno'. Error: {e:?}"
+            ))
+        })?;
+
+        let destination = get_destination(&offer).await;
+        let reply_path = match self
+            .offer_handler
+            .create_reply_path(client.clone(), self.node_id)
+            .await
+        {
+            Ok(reply_path) => reply_path,
+            Err(e) => return Err(Status::internal(format!("Internal error: {e}"))),
+        };
+
+        let info = client
+            .lightning()
+            .get_info(GetInfoRequest {})
+            .await
+            .expect("failed to get info")
+            .into_inner();
+        let network = get_network(info)
+            .await
+            .map_err(|e| Status::internal(format!("{e:?}")))?;
+
+        let cfg = PayOfferParams {
+            offer,
+            amount: inner_request.amount,
+            network,
+            client,
+            destination,
+            reply_path: Some(reply_path),
+        };
+
+        let payment = match self.offer_handler.pay_offer(cfg).await {
+            Ok(payment) => {
+                log::info!("Payment succeeded.");
+                payment
+            }
+            Err(e) => match e {
+                OfferError::AlreadyProcessing => {
+                    return Err(Status::already_exists(format!("{e}")))
+                }
+                OfferError::InvalidAmount(e) => {
+                    return Err(Status::invalid_argument(e.to_string()))
+                }
+                OfferError::InvalidCurrency => {
+                    return Err(Status::invalid_argument(format!("{e}")))
+                }
+                _ => return Err(Status::internal(format!("Internal error: {e}"))),
+            },
+        };
+
+        let reply = PayOfferResponse {
+            payment_preimage: payment.payment_preimage,
+        };
+
+        Ok(Response::new(reply))
+    }
+}
+
+// We need to check that the client passes in a tls cert pem string, hexadecimal macaroon,
+// and address, so they can connect to LND.
+fn check_auth_metadata(metadata: &MetadataMap) -> Result<LndCfg, Status> {
+    let cert_pem = match metadata.get("cert_pem") {
```

Comment:
Accepting the cert and macaroon through metadata like this is only being used as a passthrough to lnd. It's not being used to establish a secure SSL/TLS connection between the lndk gRPC client and the server. Which means we are establishing an insecure session, over which we are passing an lnd cert and macaroon for server-side processing and authorization between lndk and lnd.